src/main/java/edu/kit/compiler/optimizations/unrolling/LoopAnalysis.java

```diff
@@ -83,10 +71,66 @@ public void visitBlock(Block block) {
     }
 
     /**
-     * An exception to indicate that a graph is not reducible. See "Compilers
-     * Principles, Techniques, and Tools" (Aho et al., Chapter 9.6.4).
+     * Returns a map of each block in the graph to its nearest surrounding loop.
+     * More formally, for each block B find a loop L such that B is contained
+     * in the body of L and and there is no Loop L' such that B is contained in
+     * the body of L' and the header of L' is not contained in the body of L.
      */
-    private static final class GraphNotReducibleException extends RuntimeException {
+    private Map<Block, Loop> findSurroundingLoops(Graph graph) {
+        var surroundingLoops = new HashMap<Block, Loop>();
+
+        graph.walkBlocksPostorder(new BlockWalker() {
+            @Override
+            public void visitBlock(Block block) {
+                var loop = findSurroundingLoop(block);
+                if (loop != null) {
+                    assert loop.body.contains(block);
+                    assert loops.values().stream().noneMatch(other -> other.body.contains(block)
+                            && loop.body.contains(other.header));
+                    surroundingLoops.put(block, loop);
+                }
+            }
+        });
+
+        return surroundingLoops;
+    }
+
+    private Loop findSurroundingLoop(Block block) {
+        var dom = block.ptr;
+        while ((dom = binding_irdom.get_Block_idom(dom)) != null) {
```

Comment:
I'm not completely sure I understand why this loops starts with the `idom` of `block` instead of `block` itself. This will cause the header to be not included, right? But wouldn't including the header simplify `checkLoopsClosed` (by not requiring the extra check for the header anymore)?

Reply:
You're right that including the header would get rid of the need for that particular check. However, defining the surrounding loops as it is done here, has a number of advantages. For one thing, to walk up the hierarchy of loops, you can simply look up the surrounding loop of each loop header. It is also more consistent with the concept of a _closed_ loop, as it is perfectly fine for jump to exist from the loop header to outside the loop.

I considered both versions of the definition and found that not including the loop header to be more consistent (also intuitively speaking, a loop does not really surround its own header in my view).

Reply:
Okay, then you could maybe add a comment that explains that the header is intentionally exluded